README.md
# rack-override-path
`Rack::OverridePath` overrides responses for web applications

## Installation

Install the gem:

`gem install rack-override-path`

Or in your Gemfile:

```ruby
gem 'rack-override-path'
```

## How to use

## Example webserver setup using webrick
```
require 'rack/override-path'
require 'webrick'

app = lambda do |_env|
[200, { 'Content-Type' => 'text/plain' }, ['Hello World']]
end

Rack::Handler::WEBrick.run Rack::OverridePath.new(app)
```

### Configure Override Response
Specify a `path`, with one or more Override Parameters.

Subsequent requests matching `path` will be overridden.
#### Example
```
POST /override/path
{
  "delay": 2,
  "headers": { "Content-Type": "text/plain" },
  "status": 404,
  "body": "Nothing found",
  "method": "GET",
  "path": "/index.html"
}
```
In the example above, `GET` requests for `/index.html` will respond with a `404` status with `Nothing found` in the body after a `2` second delay

The `path` can be literal (e.g `/index.html`) or a regular expression (e.g `.*videos.*`)

#### Available Override Parameters
At least one Override Parameter must be included in each Configured Override 
* `body`
* `delay` - should be an Integer value, in seconds
* `headers`
* `status` - should be an Integer value


#### Available Filters
Filters are optional
* `method` - acceptable values are `GET`,`PUT`, `POST`, `PATCH`, `DELETE`, `HEAD`, `OPTIONS`. If `method` parameter is not specified, all methods for the matching `path` are overridden

### List Overridden Responses
#### Example
```
GET /override/path
[
    {
        "delay": 2,
        "headers": {
            "Content-Type": "text/plain"
        },
        "status": 404,
        "body": "Nothing found",
        "method": "GET",
        "path": "/index.html"
    }
]
```

### Delete all Overridden Responses
#### Example
```
DELETE /override/path
[]
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

For info on how to format all future additions to this file please reference [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2021-04-07
### Added
- Changelog reference in gem [@adambonsu](https://github.com/adambonsu).
- Delete all configured Overrides with `DELETE /override/path` [@adambonsu](https://github.com/adambonsu).

## [0.1.0] - 2021-04-07
### Added
- `method` (PUT, POST, etc) Filter to more tightly specify an override  [@adambonsu](https://github.com/adambonsu).
- Changelog [@adambonsu](https://github.com/adambonsu).
- Readme section on how to Configure an Override Response [@adambonsu](https://github.com/adambonsu).
- Readme section on Available Override Parameters [@adambonsu](https://github.com/adambonsu).
- Readme section on Available Filters [@adambonsu](https://github.com/adambonsu).


## [0.0.2] - 2021-04-02
### Added
- List Overridden Paths with `GET /override/path` [@adambonsu](https://github.com/adambonsu).

### Changed
- `path` can be either a literal or a regular expression [@adambonsu](https://github.com/adambonsu).

## [0.0.1] - 2021-04-02
### Added
- Configure Override Path with `POST /override/path` [@adambonsu](https://github.com/adambonsu).
- Override Path Gem. [@adambonsu](https://github.com/adambonsu).
- Override Parameters: `body`, `delay`, `headers`, `status` [@adambonsu](https://github.com/adambonsu).


[Unreleased]: https://github.com/adambonsu/rack-override-path/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/adambonsu/rack-override-path/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/adambonsu/rack-override-path/compare/v0.0.2...v0.1.0
[0.0.2]: https://github.com/adambonsu/rack-override-path/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/adambonsu/rack-override-path/v0.0.1

spec/flow_spec.rb
# frozen_string_literal: true

require 'rubygems'
require 'rack'
require 'rack/test'

class MockRackApp
  attr_reader :request_body

  def initialize
    @request_headers = {}
  end

  def call(env)
    @env = env
    @request_body = env['rack.input'].read
    [200, { 'Content-Type' => 'text/plain' }, ['OK']]
  end

  def [](key)
    @env[key]
  end
end

describe Rack::OverridePath do
  let(:app) { MockRackApp.new }
  subject { Rack::OverridePath.new(app) }
  let(:request) { Rack::MockRequest.new(subject) }
  let(:default_rack_environment) { Rack::MockRequest::DEFAULT_ENV }
  let(:default_app_response) { MockRackApp.new.call(default_rack_environment) }
  let(:default_app_response_status) { default_app_response[0] }
  let(:default_app_response_headers) { default_app_response[1] }
  let(:default_app_response_body) { default_app_response[2].join }

  def configure_override(override)
    response = request.post('/override/path', input: override.to_json)
    expect(response.status).to eq 200
    response
  end

  def configured_override_for(path)
    override = { 'path' => path }
    response = request.get('/override/path', input: override.to_json)
    expect(response.status).to eq 200
    response.body
  end

  context 'No override configured' do
    it 'Returns web app response' do
      response = request.get('/path/to/somewhere')
      expect(response.status).to eq default_app_response_status
      expect(response.headers).to eq default_app_response_headers
      expect(response.body).to eq default_app_response_body
    end
  end
  context 'Override configured' do
    let(:override) { { 'path' => '/index.html' } }
    context 'Delay overridden by 5 seconds' do
      let(:overridden_delay) { 5 }
      before do
        override['delay'] = overridden_delay
        configure_override(override)
      end
      it '5 second Delay Overridden for Specified Path' do
        expect(JSON.parse(configured_override_for(override['path'])).first['delay']).to eq overridden_delay
      end
      it 'Response to Overridden Request takes about 5 seconds to respond', :slow do
        time_before_request = Time.now
        _response = request.get override['path']
        time_after_request = Time.now
        expect(time_after_request - time_before_request).to be > override['delay']
      end
    end
    context 'Headers overridden' do
      context 'No Headers configured' do
        before do
          override['headers'] = {}
          configure_override(override)
        end
        it 'Override Headers is empty' do
          expect(JSON.parse(configured_override_for(override['path'])).first['headers']).to be_empty
        end
        it 'Response to Overridden Request has no Headers' do
          response = request.get override['path']
          expect(response.headers).to be_empty
        end
      end
      context 'Single Header configured' do
        before do
          override['headers'] = { 'Content-Type' => 'application/json' }
          configure_override(override)
        end
        it 'Override has one Header' do
          configured_override = JSON.parse(configured_override_for(override['path'])).first
          expect(configured_override['headers'].size).to eq 1
          expect(configured_override['headers']).to eq override['headers']
        end
        it 'Response to Overridden Request contains configured Header' do
          response = request.get override['path']
          expect(response.headers).to eq override['headers']
        end
      end
      context 'Multiple Headers configured' do
        before do
          override['headers'] = { 'Content-Type' => 'application/json', 'Content-Encoding' => 'gzip' }
          configure_override(override)
        end
        it 'Override contains configured Headers' do
          configured_override = JSON.parse(configured_override_for(override['path'])).first
          expect(configured_override['headers'].size).to eq override['headers'].size
          expect(configured_override['headers']).to eq override['headers']
        end
        it 'Response to Overridden Request contains configured Headers' do
          response = request.get override['path']
          expect(response.headers).to eq override['headers']
        end
      end
    end
    context 'Body overridden' do
      before do
        override['body'] = 'Jambo jambo'
        configure_override(override)
      end
      it 'Override contains configured Body' do
        configured_override = JSON.parse(configured_override_for(override['path'])).first
        expect(configured_override['body']).to eq override['body']
      end
      it 'Response to Overridden Request contains configured Body' do
        response = request.get override['path']
        expect(response.body).to eq override['body']
      end
    end
    context 'Status overridden' do
      before do
        override['status'] = 999
        configure_override(override)
      end
      it 'Override contains configured Status' do
        configured_override = JSON.parse(configured_override_for(override['path'])).first
        expect(configured_override['status']).to eq override['status']
      end
      it 'Response to Overridden Request contains configured Status' do
        response = request.get override['path']
        expect(response.status).to eq override['status']
      end
    end
    context 'Method overridden' do
      let(:overridden_method) { 'POST' }
      let(:overridden_status) { 808 }
      before do
        override.merge!({ 'method' => overridden_method, 'status' => overridden_status })
        configure_override(override)
      end
      it 'Override contains configured Method' do
        configured_override = JSON.parse(configured_override_for(override['path'])).first
        expect(configured_override['method']).to eq override['method']
      end
      it 'Response to Overridden Method Request is overridden' do
        response = request.send(overridden_method.downcase.to_sym, override['path'])
        expect(response.status).to eq overridden_status
      end
      it 'Non-Overridden Method Requests return App response' do
        expect(request.get(override['path']).status).to eq 200
      end
    end
    context 'Request overridden path' do
      let(:overridden_status) { 206 }
      let(:override) { { 'status' => overridden_status } }
      context 'Literal path match' do
        context 'Literal path' do
          before do
            override['path'] = '/index.html'
            response = request.post('/override/path', input: override.to_json)
            expect(response.status).to eq 200
          end

          it 'Returns overridden response' do
            response = request.get '/index.html'
            expect(response.status).to eq override['status']
          end
        end
      end
      context 'Regex path match' do
        context 'literal path' do
          before do
            override['path'] = '.*videos.*'
            response = request.post('/override/path', input: override.to_json)
            expect(response.status).to eq 200
          end

          it 'Returns overridden response' do
            response = request.get '/path/to/videos/location'
            expect(response.status).to eq override['status']
          end
        end
      end
    end
    context 'Request path that has not been overridden' do
      it 'Returns web app response' do
        response = request.get('/path/to/somewhere')
        expect(response.status).to eq default_app_response_status
        expect(response.headers).to eq default_app_response_headers
        expect(response.body).to eq default_app_response_body
      end
    end
  end
  describe 'GET /override/path' do
    context 'No Overrides configured' do
      it 'Overrides list is empty' do
        response = request.get '/override/path'
        expect(response.status).to eq 200
        expect(JSON.parse(response.body)).to be_empty
      end
    end
    context 'Override configured' do
      context 'One Override configured' do
        let(:override) { { 'status' => '999', 'path' => '.*videos.*' } }
        before do
          response = request.post '/override/path', input: override.to_json
          expect(response.status).to eq 200
        end

        it 'One Override listed' do
          response = request.get '/override/path'
          expect(response.status).to eq 200
          expect(JSON.parse(response.body)).to eq [override]
        end
      end
      context 'Multiple overrides configured' do
        let(:overrides) { [{ 'status' => '999', 'path' => '.*videos.*' }, { 'status' => '808', 'path' => '.*index.html$' }] }
        before do
          overrides.each do |override|
            response = request.post '/override/path', input: override.to_json
            expect(response.status).to eq 200
          end
        end

        it 'Overrides stacked - last override at the top' do
          response = request.get '/override/path'
          expect(response.status).to eq 200
          expect(JSON.parse(response.body)).to eq overrides.reverse
        end
      end
    end
  end
  describe 'DELETE /override/path' do
    it 'Returns status 200' do
      response = request.delete '/override/path'

      expect(response.status).to eq 200
      expect(JSON.parse(response.body)).to be_empty
    end
    context 'No Overrides configured' do
      context 'Subsequent GET /override/path' do
        it 'Overrides list is empty' do
          request.delete '/override/path'

          response = request.get '/override/path'
          expect(response.status).to eq 200
          expect(JSON.parse(response.body)).to be_empty
        end
      end
    end
    context 'Override configured' do
      let(:override) { { 'status' => '999', 'path' => '.*videos.*' } }
      before do
        response = request.post '/override/path', input: override.to_json
        expect(response.status).to eq 200

        response = request.get '/override/path'
        expect(JSON.parse(response.body).size).not_to be_zero
      end

      context 'Subsequent GET /override/path' do
        it 'Overrides list is empty' do
          request.delete '/override/path'

          response = request.get '/override/path'
          expect(response.status).to eq 200
          expect(JSON.parse(response.body)).to be_empty
        end
      end
    end
  end
end


spec/spec_helper.rb
# frozen_string_literal: true

require 'override_path'
require 'rack/test'

RSpec.configure do |config|
  config.include Rack::Test::Methods
  #
  #   # Allows RSpec to persist some state between runs in order to support
  #   # the `--only-failures` and `--next-failure` CLI options. We recommend
  #   # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = 'spec/.examples'
  #
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end
  #
  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end
  #
  #   # This setting enables warnings. It's recommended, but in some cases may
  #   # be too noisy due to issues in dependencies.
  config.warnings = true
  #
  #   # Many RSpec users commonly either run the entire suite or an individual
  #   # file, and it's useful to allow more verbose output when running an
  #   # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end
  #
  #   # Print the 10 slowest examples and example groups at the
  #   #   # end of the spec run, to help surface which specs are running
  #   #   # particularly slow.
  #   #   config.profile_examples = 10
  #   #
  #
  #   # Run specs in random order to surface order dependencies. If you find an
  #   # order dependency and want to debug it, you can fix the order by providing
  #   # the seed, which is printed after each run.
  #   #     --seed 1234
  config.order = :random
  #
  #   # Seed global randomization in this process using the `--seed` CLI option.
  #   #   # Setting this allows you to use `--seed` to deterministically reproduce
  #   #   # test failures related to randomization by passing the same `--seed` value
  #   #   # as the one that triggered the failure.
  Kernel.srand config.seed
end


spec/override_path_spec.rb
# frozen_string_literal: true

require 'rubygems'
require 'rack'
require 'rack/test'

describe Rack::OverridePath do
  let(:app) do
    app = Rack::Builder.new do
      use Rack::OverridePath
      run ->(_env) { [200, { 'Content-Type' => 'text/plain' }, ['Hello World']] }
    end
    app
  end

  context 'No override configured' do
    it 'returns web app response' do
      get '/path/to/somewhere'
      expect(last_response.status).to eq 200
      expect(last_response.body).to eq 'Hello World'
    end
  end
  describe 'POST /override/path' do
    context 'Failures' do
      context 'No body' do
        it '400 - bad config' do
          post '/override/path'
          expect(last_response.status).to eq 400
        end
      end
      context 'Empty body' do
        it '400 - bad config' do
          post '/override/path', {}
          expect(last_response.status).to eq 400
        end
      end
      context 'Body not in json format' do
        it '400 - bad config' do
          post '/override/path', 'delay' => 1, 'status' => 808, 'body' => 'la la la'
          expect(last_response.status).to eq 400
        end
      end
      context 'Body in json format' do
        context 'Body missing path parameter' do
          context 'Only delay parameter specified' do
            it '400 - bad config' do
              data = {
                'delay' => 1
              }
              post '/override/path', data.to_json, 'CONTENT_TYPE' => 'application/json'
              expect(last_response.status).to eq 400
            end
          end
          context 'Only headers parameter specified' do
            it '400 - bad config' do
              data = {
                'headers' => ['Content-Type' => 'text/plain']
              }
              post '/override/path', data.to_json, 'CONTENT_TYPE' => 'application/json'
              expect(last_response.status).to eq 400
            end
          end
          context 'Only status parameter specified' do
            it '400 - bad config' do
              data = {
                'status' => 808
              }
              post '/override/path', data.to_json, 'CONTENT_TYPE' => 'application/json'
              expect(last_response.status).to eq 400
            end
          end
          context 'Only body parameter specified' do
            it '400 - bad config' do
              data = {
                'body' => 'la la la'
              }
              post '/override/path', data.to_json, 'CONTENT_TYPE' => 'application/json'
              expect(last_response.status).to eq 400
            end
          end
        end
        context 'Body not specifying at least one override parameter' do
          it '400 - bad config' do
            data = {
              'path' => '.*videos.*'
            }
            post '/override/path', data.to_json, 'CONTENT_TYPE' => 'application/json'
            expect(last_response.status).to eq 400
          end
        end
      end
    end
    context 'Success' do
      context 'Body in json format with path and at least one override parameter' do
        context 'Path parameter specified' do
          let(:data) { { 'path' => '.*videos.*' } }
          context 'Delay parameter specified' do
            it '200 - Success' do
              data['delay'] = 1
              post '/override/path', data.to_json, 'CONTENT_TYPE' => 'application/json'
              expect(last_response.status).to eq 200
            end
          end
          context 'Headers parameter specified' do
            it '200 - Success' do
              data['headers'] = ['Content-Type' => 'text/plain']
              post '/override/path', data.to_json, 'CONTENT_TYPE' => 'application/json'
              expect(last_response.status).to eq 200
            end
          end
          context 'Status parameter specified' do
            it '200 - Success' do
              data['status'] = 808
              post '/override/path', data.to_json, 'CONTENT_TYPE' => 'application/json'
              expect(last_response.status).to eq 200
            end
          end
          context 'Body parameter specified' do
            it '200 - Success' do
              data['body'] = 'Jambo Jambo'
              post '/override/path', data.to_json, 'CONTENT_TYPE' => 'application/json'
              expect(last_response.status).to eq 200
            end
          end
        end
      end
    end
  end
end


lib/rack/override_path.rb
# frozen_string_literal: true

require 'json'
require 'rack'

module Rack
  class OverridePath
    DEFAULT_BODY = ''
    DEFAULT_DELAY = 0
    DEFAULT_HEADERS = {}.freeze
    DEFAULT_STATUS = 200
    OVERRIDE_ENDPOINT = '/override/path'
    OVERRIDE_KEYS = %w[body delay headers status].freeze

    def initialize(app)
      @app = app
      @overridden_paths = []
    end

    def call(env)
      req = request(env)
      if override_path?(req)
        override_path(req)
      elsif path_overridden?(req.path) && override_matches_method?(req)
        handle_override(req.path, env)
      elsif get_override?(req)
        get_override(req)
      elsif delete_overrides?(req)
        delete_overrides(req)
      else
        @app.call(env)
      end
    end

    def delete_overrides(req)
      @overridden_paths.clear
      get_override(req)
    end

    def delete_overrides?(req)
      override_endpoint?(req) && req.delete?
    end

    def override_matches_method?(req)
      o = override(req.path)
      o['method'].nil? ? true : o['method'].downcase == req.request_method.downcase
    end

    def override_endpoint?(req)
      req.path == '/override/path'
    end

    def get_override?(req)
      override_endpoint?(req) && req.get?
    end

    def get_override(_req)
      [200, { 'Content-Type' => 'application/json' }, [@overridden_paths.to_json]]
    end

    def override(path)
      @overridden_paths.find do |override|
        override_path = literal_path?(override['path']) ? literal_path(override['path']) : override['path']
        path.match(Regexp.new(override_path))
      end
    end

    def literal_path?(path)
      path[0] == '/'
    end

    def literal_path(path)
      "^#{path}$"
    end

    def override_path?(req)
      override_endpoint?(req) && req.post?
    end

    def override_path(req)
      payload = request_body_json(req)
      if valid?(payload)
        @overridden_paths.prepend(payload)
        override_path_successful_response(payload)
      else
        override_path_failed_response_no_body
      end
    end

    def valid?(payload)
      payload&.key?('path') && at_least_one_override_key?(payload)
    end

    def at_least_one_override_key?(payload)
      !(payload.keys & OVERRIDE_KEYS).empty?
    end

    def override_path_successful_response(payload)
      [200, { 'Content-Type' => 'application/json' }, [payload.to_json]]
    end

    def override_path_failed_response_no_body
      [400, { 'Content-Type' => 'application/json' }, [{ 'message' => 'No body provided' }.to_json]]
    end

    def path_overridden?(path)
      override(path).nil? ? false : true
    end

    def handle_override(path, env)
      o = override(path)
      handle_override_delay(o['delay']) if o['delay']
      status, headers, body = @app.call(env) unless o['body']
      handle_override_response(o, status, headers, body)
    end

    def handle_override_delay(delay)
      sleep delay
    end

    def handle_override_response(override, status, headers, body)
      response_status = override['status'] || status || DEFAULT_STATUS
      response_headers = override['headers'] || headers || DEFAULT_HEADERS
      response_body = override['body'] || body || DEFAULT_BODY
      [response_status, response_headers, prepare_response_body_for_rack(response_body)]
    end

    def prepare_response_body_for_rack(response_body)
      response_body = response_body.is_a?(Hash) ? response_body.to_json : response_body
      response_body.is_a?(Array) ? response_body : [response_body]
    end

    def request(env)
      Rack::Request.new(env)
    end

    def request_body_json(req)
      JSON.parse(request_body(req))
    rescue StandardError => _e
      nil
    end

    def request_body(req)
      req.body.rewind
      req.body.read
    rescue StandardError => _e
      ''
    end
  end
end
